core/trino-main/src/main/java/io/trino/sql/planner/sanity/ValidateLimitWithPresortedInput.java

```diff
@@ -0,0 +1,114 @@
+import static io.trino.sql.planner.optimizations.StreamPropertyDerivations.derivePropertiesRecursively;
+
+/**
+ * Verifies that input of order-sensitive Limit is ordered by the pre-sorted symbols
+ */
+public class ValidateLimitWithPresortedInput
+        implements Checker
+{
+    @Override
+    public void validate(PlanNode planNode,
+            Session session,
+            Metadata metadata,
+            TypeOperators typeOperators,
+            TypeAnalyzer typeAnalyzer,
+            TypeProvider types,
+            WarningCollector warningCollector)
+    {
+        planNode.accept(new Visitor(session, metadata, typeOperators, typeAnalyzer, types), null);
+    }
+
+    private static final class Visitor
+            extends PlanVisitor<Void, Void>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final TypeOperators typeOperators;
+        private final TypeAnalyzer typeAnalyzer;
+        private final TypeProvider types;
+
+        private Visitor(Session session,
+                Metadata metadata,
+                TypeOperators typeOperators,
+                TypeAnalyzer typeAnalyzer,
+                TypeProvider types)
+        {
+            this.session = session;
+            this.metadata = metadata;
+            this.typeOperators = typeOperators;
+            this.typeAnalyzer = typeAnalyzer;
+            this.types = types;
+        }
+
+        @Override
+        protected Void visitPlan(PlanNode node, Void context)
+        {
+            node.getSources().forEach(source -> source.accept(this, context));
+            return null;
+        }
+
+        @Override
+        public Void visitLimit(LimitNode node, Void context)
+        {
+            PlanNode source = node.getSource();
+            source.accept(this, context); // visit child
+
+            if (node.getPreSortedInputs().isEmpty()) {
+                return null;
+            }
+
+            StreamProperties properties = derivePropertiesRecursively(node.getSource(), metadata, typeOperators, session, types, typeAnalyzer);
+
+            // We do not use LocalProperties#match to fully verify sorting properties
+            // as OrderingScheme of input is not tracked by LimitNode
+            List<Symbol> sortedColumns = normalizeAndPrune(properties.getLocalProperties()).stream()
+                    .filter(property -> property instanceof SortingProperty)
+                    .flatMap(property -> property.getColumns().stream())
+                    .collect(toImmutableList());
+
+            PeekingIterator<Symbol> sortedColumnsIterator = peekingIterator(sortedColumns.iterator());
+            for (Symbol symbol : node.getPreSortedInputs()) {
+                if (!sortedColumnsIterator.hasNext() || !sortedColumnsIterator.next().equals(symbol)) {
+                    throw new VerifyException("Limit with pre-sorted inputs is not sorted by " + symbol);
+                }
+            }
```

Comment:
This can be simplified to:

```java
if (sortedColumns.size() > node.getPreSortedInputs().size() ||
        !sortedColumns.subList(0, node.getPreSortedInputs().size()).equals(node.getPreSortedInputs())) {
    throw new VerifyException(format("Expected Limit input to be sorted by: %s, but was %s" + node.getPreSortedInputs(), sortedColumns));
}
```

Also, I think it's important to show all the columns it expects to be sorted on, as the order in which they are listed matters.

Reply:
done